icu4c/source/i18n/tibetancal.cpp

```diff
@@ -0,0 +1,538 @@
+//-------------------------------------------------------------------------
+
+
+TibetanCalendar* TibetanCalendar::clone() const {
+  return new TibetanCalendar(*this);
+}
+
+TibetanCalendar::TibetanCalendar(const Locale& aLocale, UErrorCode& success)
+  :   Calendar(TimeZone::forLocaleOrDefault(aLocale), aLocale, success)
+{
+  setTimeInMillis(getNow(), success); // Call this again now that the vtable is set up properly.
+}
+
+TibetanCalendar::TibetanCalendar(const TibetanCalendar& other) : Calendar(other) {
+}
+
+TibetanCalendar::~TibetanCalendar()
+{
+}
+
+/**
+ * return the calendar type, "tibetan" or "tibetan-tsurphu".
+ *
+ * @return calendar type
+ * @internal
+ */
+const char *TibetanCalendar::getType() const { 
+   const char *sType = NULL;
+
+   switch (cType) {
+    case PHUGPA:
+      sType = 'tibetan';
+      break;
+    case TSURPHU:
+      sType = 'tibetan-tsurphu';
+      break;
+    default:
+      UPRV_UNREACHABLE; // out of range
+   }
+   return sType;
+}
+
+
+/**
+ * Sets Tibetan calendar calculation type used by this instance.
+ *
+ * @param type    The calendar calculation type, <code>PHUGPA</code> to use the phugpa
+ *                calendar, <code>TSURPHU</code> to use the tsurphu calendar.
+ * @internal
+ */
+void TibetanCalendar::setCalculationType(ECalculationType type, UErrorCode &status)
+{
+    if (cType != type) {
+        // The fields of the calendar will become invalid, because the calendar
+        // rules are different
+        UDate m = getTimeInMillis(status);
+        cType = type;
+        clear();
+        setTimeInMillis(m, status);
+    }
+}
+
+
+/**
+* Returns <code>true</code> if this object is using the phugpa
+* calendar, or <code>false</code> if using the tsurphu calendar.
+*/
+UBool TibetanCalendar::isPhugpa() {
+    return (cType == PHUGPA);
+}
+
+
+
+//-------------------------------------------------------------------------
+// Assorted calculation utilities
+//-------------------------------------------------------------------------
+
+// fixed tables
+int32_t SUN_TAB[] = {0, 6, 10, 11};
+int32_t MOON_TAB[] = {0, 5, 10, 15, 19, 22, 24, 25};
+
+// mean date
+double PHUGPA_M0 = 2015501 + 4783.0/5656;
+double TSURPHU_M0 = 2353745 + 1795153 / 7635600;
+double M1 = 167025.0/5656;
+double M2 = M1/30;
+
+// mean sun
+double PHUGPA_S0 = 743.0/804;
+double TSURPHU_S0 = -5983/108540;
+double S1 = 65.0/804;
+double S2 = S1/30;
+
+// anomaly moon
+double PHUGPA_A0 = 475.0/3528;
+double TSURPHU_A0 = 207 / 392;;
+double A1 = 253.0/3528;
+double A2 = 1.0/28;
+
+int32_t PHUGPA_YEAR0 = 806;
+int32_t PHUGPA_MONTH0 = 3;
+int32_t PHUGPA_BETA_STAR = 61;
+double PHUGPA_ALPHA = 1 + 827.0/1005;
+
+int32_t TSURPHU_YEAR0 = 1732;
+int32_t TSURPHU_MONTH0 = 3;
+int32_t TSURPHU_BETA_STAR = 59;
+double TSURPHU_ALPHA = 2 + 1052/9045;
+
+/**
+ * Return the moonTab by using the table for the given integer value. 
+ * @param moonTab An integer number
+ */
+int32_t TibetanCalendar::moonTab(int32_t moonTab) const {
```

Comment:
Should the parameter name rename to i to improve readability while reference to (7.18) in [Janson] ?
The comments need improvement. It does not make sense to me that you pass in a moonTab and return a modified moonTab. 
nit: Return -> Returns.
What is "moon tab" anyway?

Reply:
see my eariler feedback around MOON_TAB. 
1. I think that could be just a static constexpr inside the function
2. the function could be just a simple internal C style internal function and has no need to be part of the interface as a member method
3. we can avoid the complex if check by just unfold the array to 12 and 28 elements which the increase the table by 8x4=32 bytes 20x4=80 bytes but probably reduce the compiled code. If we really concern of the size increase, we can actually make the constexpr int8_t instead but I think that may not be necessary.